### PR TITLE
Injective STM and Async type families

### PR DESCRIPTION
**Pull request by coot**
- io-classes: use injective type families
- io-sim: updated
- Updated network-mux, ouroboros-network-framework and ouroboros-network
- Updated ouroboros-consensus and ouroboros-consensus-test

### CODE REVIEW

File: ouroboros-consensus-test/test-consensus/Test/Consensus/BlockchainTime/Simple.hs [hunk 1]
```diff
@@ -373,6 +374,100 @@ instance MonadDelay (OverrideDelay (IOSim s)) where
...
+deriving instance MonadPlus   (STM m) => MonadPlus   (OverrideDelaySTM m)
+
+instance (MonadSTM m, MonadThrow (STM m)) => MonadThrow (OverrideDelaySTM m) where
+    throwIO = OverrideDelaySTM . lift . throwIO
+    bracket before after f =
+         OverrideDelaySTM $ ReaderT $ \schedule ->
+          bracket       (runReaderT (unOverrideDelaySTM  before)   schedule)
+                  (\a -> runReaderT (unOverrideDelaySTM (after a)) schedule)
+                  (\a -> runReaderT (unOverrideDelaySTM (f a))     schedule)
+
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-consensus/Test/Consensus/BlockchainTime/Simple.hs:394]: Style: There's an extraneous newline here.

File: ouroboros-consensus-test/test-consensus/Test/Consensus/BlockchainTime/Simple.hs [hunk 2]
```diff
@@ -373,6 +374,100 @@ instance MonadDelay (OverrideDelay (IOSim s)) where
...
+            let unmask' :: forall x. OverrideDelay m x -> OverrideDelay m x
+                unmask' (OverrideDelay (ReaderT m)) =
+                  OverrideDelay $ ReaderT $ \schedule' -> unmask (m schedule')
+            in runReaderT (unOverrideDelay $ f unmask') schedule
+        )
+
+  waitCatchSTM = OverrideDelaySTM . lift . waitCatchSTM . unOverrideDelayAsync
+  pollSTM      = OverrideDelaySTM . lift . pollSTM . unOverrideDelayAsync
+
+
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-consensus/Test/Consensus/BlockchainTime/Simple.hs:470]: Style: There's an extraneous newline here.

**Review by nfrisby — COMMENTED** [review]
I Approve the `ouroboros-consensus` parts. There are some style hiccups, but it's nothing consequential and it's limited to a testing module.

Thanks much for updating the Consensus code, Marcin.

File: ouroboros-consensus-test/test-consensus/Test/Consensus/BlockchainTime/Simple.hs (hunk 1 above)
Comment (coot) [reply ouroboros-consensus-test/test-consensus/Test/Consensus/BlockchainTime/Simple.hs:394]: ok I'll change that, we often use 2 new lines as a break in `ouroboros-network`.

File: ouroboros-consensus-test/test-consensus/Test/Consensus/BlockchainTime/Simple.hs (hunk 2 above)
Comment (coot) [reply ouroboros-consensus-test/test-consensus/Test/Consensus/BlockchainTime/Simple.hs:470]: fixed

**Review by bolt12 — APPROVED** [review]
LGTM

**Comment by coot** [comment]
bors merge

**Comment by coot** [comment]
bors merge

**Comment by coot** [comment]
bors merge

**Comment by coot** [comment]
bors merge

**Pull request merged**